File: src/main/java/com/hoi4utils/clausewitz/data/tooltip/CustomTooltip.java
package com.hoi4utils.clausewitz.data.tooltip;

import com.hoi4utils.clausewitz.localization.Localization;
import com.hoi4utils.clausewitz_parser.Node;
import com.hoi4utils.clausewitz_parser.Parser;
import com.hoi4utils.clausewitz_parser.ParserException;
import scala.jdk.javaapi.CollectionConverters;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

/*
 * CustomTooltip File
 */
public class CustomTooltip {
	String tooltipID;
	public static ArrayList<CustomTooltip> tooltips;
	private Localization localization;      // localization text for tooltip id

	public CustomTooltip(String ID) {
		this.tooltipID = ID;
	}

	public static List<Function<CustomTooltip, ?>> getDataFunctions() {
		List<Function<CustomTooltip, ?>> dataFunctions = new ArrayList<>(2);         // 2 for optimization, limited number of data functions.

		dataFunctions.add(CustomTooltip::id);
		dataFunctions.add(CustomTooltip::getTooltipText);

		return dataFunctions;
	}

	public String id() {
		return tooltipID;
	}

	public static void loadTooltips(File file) {
		tooltips = new ArrayList<>();

		if (!validFile(file)) return;

		Parser parser = new Parser(file);
		Node rootNode = null;
		try {
			rootNode = parser.parse();
		} catch (ParserException e) {
			throw new RuntimeException(e);
		}
		//Node[] tooltipExpressions = rootNode.filter("custom_trigger_tooltip").toList().toArray(new Node[0]); // prev2 = prev = parser.expression().getAll("custom_trigger_tooltip");
		List<Node> tooltipExpressions = CollectionConverters.asJava(rootNode.filter("custom_trigger_tooltip").toList()); 
		for (Node exp : tooltipExpressions) {
			if (!exp.contains("tooltip")) {
				continue;
			}
			String expID = exp.getValue("tooltip").string();
			if (expID == null) {
				continue;
			}
			tooltips.add(new CustomTooltip(expID));
		}
	}

	private static boolean validFile(File file) {
		if (file == null) {
			System.err.println("File null in: " + CustomTooltip.class + "-> loadTooltips()");
			return false;
		}
		if (!file.exists()) {
			System.err.println("File does not exist: " + file);
			return false;
		}
		if (!file.isFile()) {
			System.err.println("File is not file: " + file);
			return false;
		}
		return true;
	}

	public static ArrayList<CustomTooltip> getTooltips() {
		if (tooltips.isEmpty()) {
			return null;
		}

		return tooltips;
	}

	public String toString() {
		if (tooltipID != null) {
			return tooltipID;
		}

		return super.toString();
	}

	public void setLocalization(Localization tooltipLocalization) {
		this.localization = tooltipLocalization;
	}

	public Localization localization() {
		return localization;
	}

	public String getTooltipText() {
		if (localization == null) {
			return null;
		}
		return localization.text();
	}
}

File: src/main/scala/com/hoi4utils/hoi4/tooltip/CustomTooltip.scala
package com.hoi4utils.hoi4.tooltip

import com.hoi4utils.hoi4.localization.Localization
import com.hoi4utils.parser.{Node, NodeSeq, Parser, ParserException, SeqNode}

import java.io.File
import scala.collection.mutable.ListBuffer

/**
 * Represents one custom tooltip (by its ID), plus optional localization.
 */
class CustomTooltip(val id: String) {
  private var _loc: Option[Localization] = None

  /** Raw tooltip ID (the key). */
  def tooltipID: String = id

  /** Attach the Localization record for this tooltip. */
  def setLocalization(loc: Localization): Unit =
    _loc = Option(loc)

  /** The Localization, if any, for this tooltip. */
  def localization: Option[Localization] = _loc

  /** The resolved text, if we have a Localization attached. */
  def text: Option[String] = _loc.map(_.text)

  override def toString: String = Option(id).getOrElse(super.toString)
}

object CustomTooltip {
  private var tooltips: ListBuffer[CustomTooltip] = ListBuffer.empty

  /**
   * Two data‐accessors: the ID and the resolved text.
   * Replace `Any` with a more precise type if desired.
   */
  def dataFunctions(): Iterable[CustomTooltip => ?] = List(
    _.tooltipID,
    _.text.orNull
  )

  /** Load all `custom_trigger_tooltip` nodes from the given file. */
  def loadTooltips(file: File): Unit = {
    // TODO TODO
    ()
//    tooltips = ListBuffer.empty[CustomTooltip]
//
//    if (!validFile(file)) return
//
//    val rootNode: Node[?] = try {
//      new Parser(file).parse
//    } catch {
//      case e: ParserException => throw new RuntimeException(e)
//    }
//
//    // collect all tooltip IDs
//    val expressions: NodeSeq =
//      rootNode.filter("custom_trigger_tooltip")
//
//    for {
//      exp   <- expressions
//      if exp.contains("tooltip")                   // only those that define a tooltip
//      expID <- Option(exp.getValue("tooltip").$stringOrElse(null))
//    } tooltips += new CustomTooltip(expID)
  }

  private def validFile(f: File): Boolean =
    Option(f).exists(f => f.exists && f.isFile)

  def getTooltips: Iterable[CustomTooltip] =
    if (tooltips.nonEmpty) tooltips else Nil
}
